bse/driver-jack.cc

```diff
@@ -0,0 +1,905 @@
+          input_ports_.push_back (port);
+        else
+          error = Bse::Error::FILE_OPEN_FAILED;
+
+        snprintf (port_name, port_name_size, "out_%u", i);
+        port = jack_port_register (jack_client_, port_name, JACK_DEFAULT_AUDIO_TYPE, JackPortIsOutput, 0);
+        if (port)
+          output_ports_.push_back (port);
+        else
+          error = Bse::Error::FILE_OPEN_FAILED;
+      }
+
+    /* initialize ring buffers */
+    if (error == 0)
+      {
+        // keep at least two jack callback sizes for dropout free audio
+        uint min_buffer_frames = jack_get_buffer_size (jack_client_) * 2;
+
+        // keep an extra engine buffer size (this compensates also for cases where the
+        // engine buffer size is not a 2^N value, which would otherwise cause the
+        // buffer never to be fully filled with 2 periods of data)
+        min_buffer_frames += config.block_length;
+
+        // honor the user defined latency specification
+        //
+        // the user defined latency is only used to adjust our local buffering
+        // -> it doesn't take into account latencies outside beast, such as the buffering
+        //    jack does, or latencies added by other clients)
+        uint user_buffer_frames = config.latency_ms * config.mix_freq / 1000;
+        uint buffer_frames = std::max (min_buffer_frames, user_buffer_frames);
+
+        input_ringbuffer_.resize (buffer_frames, n_channels_);
+        output_ringbuffer_.resize (buffer_frames, n_channels_);
+        buffer_frames_  = output_ringbuffer_.get_writable_frames();
+
+        // the ringbuffer should be exactly as big as requested
+        if (buffer_frames_ != buffer_frames)
+          {
+            Bse::warning ("JACK driver: ring buffer size not correct: (buffer_frames_ != buffer_frames); (%u != %u)\n",
+                          buffer_frames_, buffer_frames);
+            error = Bse::Error::INTERNAL;
+          }
+        JDEBUG ("ringbuffer size = %.3fms", buffer_frames_ / double (mix_freq_) * 1000);
+
+        /* initialize output ringbuffer with silence
+         * this will prevent dropouts at initialization, when no data is there at all
+         */
+        vector<float>	  silence (output_ringbuffer_.get_total_n_frames());
+        vector<const float *> silence_buffers (output_ringbuffer_.get_n_channels());
+
+        fill (silence_buffers.begin(), silence_buffers.end(), &silence[0]);
+
+        uint frames_written = output_ringbuffer_.write (buffer_frames, &silence_buffers[0]);
+        if (frames_written != buffer_frames)
+          Bse::warning ("JACK driver: output silence init failed: (frames_written != jack->buffer_frames)\n");
+
+      }
+
+    /* activate */
+    if (error == 0)
+      {
+        jack_set_process_callback (jack_client_, static_process_callback, this);
+        jack_set_latency_callback (jack_client_, static_latency_callback, this);
+        jack_on_shutdown (jack_client_, static_shutdown_callback, this);
```

Comment:
Note that lambdas that don't take context arguments boil down to normal functions that can be used as function pointer. For example, here is how it looks when you replace the static_* wrappers with lambda:

    jack_on_shutdown (jack_client_, [] (void *p) {
      static_cast<JackPcmDriver*> (p)->shutdown_callback();
    }, this); 

This could also be a one liner if you have a wider terminal.

Reply:
Fixed. Interesting, didn't know that this would work.